src/config.rs

```diff
@@ -0,0 +1,372 @@
+use crate::{Result, Target, Toml};
+
+use std::collections::HashMap;
+use std::env::var;
+#[derive(Debug)]
+struct Environment(&'static str, Option<HashMap<&'static str, &'static str>>);
+
+impl Environment {
+    fn new() -> Self {
+        Environment("CROSS", None)
+    }
+
+    #[cfg(test)]
+    /// for test set a map with values to mock process env
+    fn new_with(map: HashMap<&'static str, &'static str>) -> Self {
+        Environment("CROSS", Some(map))
+    }
+
+    fn build_var_name(&self, name: &str) -> String {
+        format!("{}_{}", self.0, name.to_ascii_uppercase().replace("-", "_"))
+    }
+
+    #[cfg(not(test))]
+    /// get value from process env
+    fn get_var(&self, name: &str) -> Option<String> {
+        var(name).ok().and_then(|v| Some(v))
+    }
+
+    #[cfg(test)]
+    /// for tests get value from internal map
+    fn get_var(&self, name: &str) -> Option<String> {
+        self.1
+            .as_ref()
+            .and_then(|map| map.get(name).and_then(|v| Some(v.to_string())))
+    }
+
+    fn target_path(target: &Target, key: &str) -> String {
+        format!("TARGET_{}_{}", target.triple(), key)
+    }
+
+    fn build_path(key: &str) -> String {
+        format!("BUILD_{}", key)
+    }
+
+    fn get_build_var(&self, key: &str) -> Option<String> {
+        self.get_var(&self.build_var_name(&Environment::build_path(key)))
+    }
+
+    fn get_target_var(&self, target: &Target, key: &str) -> Option<String> {
+        self.get_var(&self.build_var_name(&Environment::target_path(target, key)))
+    }
+
+    fn xargo(&self, target: &Target) -> Result<(Option<bool>, Option<bool>)> {
+        let (build_xargo, target_xargo) = (
+            self.get_build_var("XARGO"),
+            self.get_target_var(target, "XARGO"),
+        );
+        let build_env = if let Some(build) = build_xargo {
+            Some(build.parse::<bool>().or_else(|_| {
+                Err(format!(
+                    "error parsing {} from XARGO environment variable",
+                    build
+                ))
+            })?)
+        } else {
+            None
+        };
+        let target_env = if let Some(t) = target_xargo {
+            Some(t.parse::<bool>().or_else(|_| {
+                Err(format!(
+                    "error parsing {} from XARGO environment variable",
+                    t
+                ))
+            })?)
+        } else {
+            None
+        };
+
+        Ok((build_env, target_env))
+    }
+
+    fn image(&self, target: &Target) -> Option<String> {
+        self.get_target_var(target, "IMAGE")
+    }
+
+    fn runner(&self, target: &Target) -> Option<String> {
+        self.get_target_var(target, "RUNNER")
+    }
+
+    fn passthrough(&self, target: &Target) -> (Option<Vec<String>>, Option<Vec<String>>) {
+        let passthrough_target = self
+            .get_target_var(target, "env_passthrough")
+            .and_then(|ref s| Some(split_to_cloned_by_ws(s)));
```

Comment:
Use `map` for these to avoid the need for `Some`  the closure.

Reply:
changed it thanks :)